### PR TITLE
[IndexTable] prevents right-aligned header sort icon flicker on hover after selection

### PR DESCRIPTION
**Pull request by rcaplanshopify**
### WHY are these changes introduced?

tl;dr These changes add the final polish to the right-aligned table header updates.

Conditionally applying the class with the right-aligned sort icon fade out animation triggered a flash of the sort icon when the user clicked on a single row or the "select all" checkbox, then clicked on a sortable table header other than the impacted right-aligned one(s). Because the keyframe animation starts with the sort icon visible, the initial opacity value is `1.0`, and that's where the flicker originated.

I was able to apply the default sort icon fade transition to the right-aligned header on hover-end only. This eliminates the flicker and makes the sort icon fade consistent and agnostic to the order of user events (sequence of sort and selection clicks).

The logic to detect a render after a selection is still needed to prevent the sort button's slide animation from firing.

Additionally, I adjusted the left position calculation of the right-aligned sort button's `::before` pseudo-element to ensure consistency. The first iteration placement was relative to the size of the button, but the updated version is offset from the right edge of the button.

#### Before

https://user-images.githubusercontent.com/100380574/227277947-a402c4a4-8b54-4ffd-8e66-5c9086de24ac.mov

#### After

https://user-images.githubusercontent.com/100380574/227343185-ffc9d693-1e6f-4450-a2e8-484b09d7d02c.mov

Fixes #0000 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Review by mrcthms — APPROVED** [review]
Love changes that remove more then they add, video is looking really good too!
